vcr/statuslist/store_test.go

```diff
@@ -0,0 +1,256 @@
+
+			assert.NoError(t, err)
+			require.NotNil(t, entry)
+
+			assert.Equal(t, statusListCredential, entry.StatusListCredential)
+			assert.Equal(t, "0", entry.StatusListIndex)
+			assert.Equal(t, fmt.Sprintf("%s#0", statusListCredential), entry.ID)
+			assert.Equal(t, credential.StatusList2021EntryType, entry.Type)
+			assert.Equal(t, StatusPurposeRevocation, entry.StatusPurpose)
+		})
+		t.Run("second entry", func(t *testing.T) {
+			statusListCredential, _ := toStatusListCredential(aliceDID, 1)
+
+			entry, err = s.Create(nil, aliceDID, StatusPurposeRevocation)
+
+			assert.NoError(t, err)
+			require.NotNil(t, entry)
+
+			assert.Equal(t, statusListCredential, entry.StatusListCredential)
+			assert.Equal(t, "1", entry.StatusListIndex)
+			assert.Equal(t, fmt.Sprintf("%s#1", statusListCredential), entry.ID)
+
+		})
+		t.Run("credential rollover", func(t *testing.T) {
+			statusListCredential, _ := toStatusListCredential(aliceDID, 1)
+			// set last_issued_index to max value for a single credential so the next entry will be in page 2
+			s.db.Model(&statusListCredentialRecord{}).
+				Where("status_list_credential = ?", statusListCredential).
+				Update("last_issued_index", statuslist2021.MaxBitstringIndex)
+			statusListCredential, _ = toStatusListCredential(aliceDID, 2) // now expect page 2 to be used
+
+			entry, err = s.Create(nil, aliceDID, StatusPurposeRevocation)
+
+			assert.NoError(t, err)
+			require.NotNil(t, entry)
+
+			assert.Equal(t, statusListCredential, entry.StatusListCredential)
+			assert.Equal(t, "0", entry.StatusListIndex)
+			assert.Equal(t, fmt.Sprintf("%s#0", statusListCredential), entry.ID)
+		})
+		t.Run("second issuer", func(t *testing.T) {
+			entry, err = s.Create(nil, bobDID, StatusPurposeRevocation)
+
+			assert.NoError(t, err)
+			require.NotNil(t, entry)
+			assert.Equal(t, "0", entry.StatusListIndex)
+			// alice#1, alice#2, bob#1; fails if only part of the 'ok' block is executed
+			assert.Equal(t, int64(3), s.db.Find(&[]statusListCredentialRecord{}).RowsAffected)
+		})
+	})
+	t.Run("error - unsupported purpose", func(t *testing.T) {
+		entry, err = s.Create(nil, aliceDID, statusPurposeSuspension)
+		assert.ErrorIs(t, err, errUnsupportedPurpose)
+		assert.Nil(t, entry)
+	})
+	t.Run("error - unsupported ID method", func(t *testing.T) {
+		entry, err = s.Create(nil, did.MustParseDID("did:nuts:123"), StatusPurposeRevocation)
+		assert.EqualError(t, err, "status list: unsupported DID method: nuts")
+		assert.Nil(t, entry)
+	})
+	t.Run("optimistic locking", func(t *testing.T) {
+		t.Run("gorm translates errors", func(t *testing.T) {
+			// without this the lock conditions are not handled correctly
+			assert.True(t, s.db.Config.TranslateError)
+		})
+		t.Run("trigger gorm.ErrRecordNotFound on UPDATE of LastIssuedIndex", func(t *testing.T) {
+			t.SkipNow()
+			// This test is skipped since in needs manual evaluation.
+			// Run this test if the logic around optimistic locking has been changed.
```

Comment:
You can evaluate this test automatically, if you query the database for `LastIssuedIndex`? The index should be `len(jobs) - 1`? If not, a race condition occured and the idnex wasn't updated properly.

For bonus points, lower the `MaxBitStringIndex` in the test so you can also automatically test a full page (without mkaing the test super long-running).

Reply:
The test is aiming to confirm there were 2 parallel updates and that it's handled correctly. The test you propose does not guarantee the race condition occurs. (CI only has 1 core and might never trigger this situation)

Reply:
The suggested test revealed that optimistic locking does not work. Simultaneous reads of the old value during update of last_issued_index results in the next index value being issue twice. Optimistic locking is replaced with SELECT FOR UPDATE and a mutex for cases where the db is SQLite.